### PR TITLE
remote debug server (zdbserver)

### PR DESCRIPTION
**Pull request by kzsnow**

### CODE REVIEW

File: src/zelos/tools/zdbserver/test_zdbserver.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+
+            # Test `zdbserver` exceptions
+            # ...
+
+            # Test memory read after break
+            stack_var = int(zdb.read_register("esp"), 16) + 0x1C
+            data = zdb.read_memory(f"0x{stack_var:x}", 4)
+            print(data.data)
+            # self.assertEqual(data.data, b"\xc0\x00\x00\x90")  # after inst
+            self.assertEqual(data.data, b"\x80\x00\x00\x90")  # before inst
```
Comment (kzsnow) [inline src/zelos/tools/zdbserver/test_zdbserver.py:159]: Interesting. This test fails only if this test script resides in the zdbserver directory. Same test passes if in the tests/tools directory.

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+            True if no exception occurred.
+        """
+        logger.debug(f"[debug] write_register('{register}', {value})")
+        value = int(value, 0)
+        register = register.lower()
+        if register == "pc":
+            self.z.regs.setIP(value)
+        else:
+            self.z.regs[register] = value
+        return True
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:117]: Similarly to the above comment, why not return the numeric value written to the register? To be consistent with the `set_reg` function & related apis in the `RegsApi`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    def get_mappings(self) -> list:
+        """
+        Gets the memory region mappings.
+
+        Returns:
+            A list of the current memory regions.
+        """
+        logger.debug(f"[debug] get_mappings()")
+        vmmap = []
+        regions = self.z.internal_engine.memory.get_regions()
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:200]: Can replace with updated memory API:

`regions = self.z.memory.get_regions()`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        Args:
+            register: String representation of the register name.
+
+        Returns:
+            Value of the specified register.
+        """
+        logger.debug(f"[debug] read_register('{register}')")
+        register = register.lower()
+        if register == "pc":
+            return "0x{0:x}".format(self.z.regs.getIP())
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:96]: `f"0x{self.z.regs.getIP():x}"`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        return "0x{0:x}".format(self.z.regs[register])
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:97]: `f"0x{self.z.regs[register]:x}"`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+                }
+
+        """
+        logger.debug(f"[debug] run()")
+        break_state = self.z.start()
+        # Format break_state for RPC by converting `None` and Numbers to
+        # RPC-serializable values.
+        if break_state is None:
+            return {}
+        break_state["pc"] = "0x{0:x}".format(break_state["pc"])
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:239]: `break_state["pc"] = f"0x{break_state['pc']:x}"`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        if break_state.get("syscall", None) is None:
+            break_state["syscall"] = {}
+        else:
+            break_state["syscall"]["retval"] = "0x{0:x}".format(
+                break_state["syscall"]["retval"]
+            )
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:245]: `break_state["syscall"]["retval"] = f"0x{break_state['syscall']['retval']:x}"`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+            for arg in break_state["syscall"]["args"]:
+                arg["value"] = "0x{0:x}".format(arg["value"])
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:247]: `arg["value"] = f"0x{arg['value']:x}"`

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+        return True
+
+    def get_filepath(self) -> str:
+        """
+        Get the file path of the emulated binary.
+
+        Returns:
+            The full path to the emulated binary.
+        """
+        return self.z.internal_engine.main_module_name
```
Comment (rcourt) [inline src/zelos/tools/zdbserver/zdbserver.py:298]: After #58 is merged, you can replace with the updated API

`return self.z.main_module_path`

File: src/zelos/tools/zdbserver/test_zdbserver.py (hunk 1 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/test_zdbserver.py:159]: It was the differing host path affecting emulated execution. Fixed by ensuring virtual path is used in Linux emulation.

File: src/zelos/tools/zdbserver/zdbserver.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+            address: Address to start writing data to.
+            value: Binary data to write in memory.
+
+        Returns:
+            True if no exception occurred.
+        """
+        address = int(address, 0)
+        logger.debug(f"[debug] write_memory(0x{address:x}, {value.data})")
+        self.z.memory.write(address, value.data)
+        return True
```
[the remark this answers is not in the document]
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:81]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 2 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:117]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 6 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:239]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 3 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:200]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 4 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:96]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 5 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:97]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 8 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:247]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 7 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py:245]: done

File: src/zelos/tools/zdbserver/zdbserver.py (hunk 9 above)
Comment (kzsnow) [reply src/zelos/tools/zdbserver/zdbserver.py]: done

File: src/zelos/tools/zdbserver/__main__.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+# ======================================================================
+
+if __name__ == "__main__":
+    import sys
+
+    from .zdbserver import create_server
+
+    server = create_server(sys.argv[1:])
```
[the remark this answers is not in the document]
Comment (kzsnow) [reply src/zelos/tools/zdbserver/__main__.py:23]: it will be done in create_server

File: src/zelos/ext/platforms/linux/loader.py [hunk 12]
```diff
@@ -214,7 +214,8 @@ def _stack_setup(self, thread):
         #   https://github.com/torvalds/linux/blob/v3.19/include/uapi/linux/auxvec.h
 
         module_path_ptr, str_len = self.memory.heap.allocstr(
-            f"/home/admin/{self.main_module_name}", alloc_name="Module Path"
+            f"{self._z.config.virtual_path}/{self._z.config.virtual_filename}",
```
Comment (rcourt) [inline src/zelos/ext/platforms/linux/loader.py:217]: Virtual path may already include a forward slash, so you should instead use python's `posixpath.join` function to concat the `virtual_path` and `virtual_filename`

Comment (kzsnow) [reply src/zelos/ext/platforms/linux/loader.py]: done

**Pull request merged**